lib/irb/cmd/show_source.rb

```diff
@@ -45,13 +45,24 @@ def execute(str = nil)
       private
 
       def show_source(source)
-        file_content = IRB::Color.colorize_code(File.read(source.file))
-        code = file_content.lines[(source.first_line - 1)...source.last_line].join
-        content = <<~CONTENT
+        if source.file_content
+          # Colorizing partial code `source.content` sometimes fails.
+          # Instead, we need to colorize `source.file_content` and extract the relevant lines.
+          file_content = IRB::Color.colorize_code(source.file_content)
+          code = file_content.lines[(source.first_line - 1)...source.last_line].join
+        elsif source.content
+          code = IRB::Color.colorize_code(source.content)
```

Comment:
Just an idea: how about we make `Source` a normal Ruby class and encapsulate these conditions 2 conditions inside a `colorized_content` method.

And then we can add a unit test for it to cover the partial coloring failure case too.